File: Applied Course/4.Problem Solving/1.Problems on Array/24.The K Weakest Rows in a Matrix.cpp
#include<bits/stdc++.h>
using namespace std;
// Logic-1
class Solution {
public:
    vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
        int r=mat.size(),c=mat[0].size();
        vector<pair<int,int> > v;
        vector<int> result;
        for(int i=0;i<r;i++)
        {
            int count=accumulate(mat[i].begin(),mat[i].end(),0);
            v.push_back({count,i});
        }
        sort(v.begin(),v.end());
        for(int i=0;i<k;i++)
            result.push_back(v[i].second);
        return result;
    }
};
// Logic-2
class Solution {
public:
    int binarySearch(vector<int>& row)
    {
        int low=0,high=row.size();
        while(low<high)
        {
            int mid=low+(high-low)/2;
            if(row[mid]==1)
                low=mid+1;
            else
                high=mid;
        }
        return low;
    }
    vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
        int r=mat.size(),c=mat[0].size();
        vector<pair<int,int> > v;
        vector<int> result;
        for(int i=0;i<r;i++)
        {
            int count=binarySearch(mat[i]);
            v.push_back({count,i});
        }
        sort(v.begin(),v.end());
        for(int i=0;i<k;i++)
            result.push_back(v[i].second);
        return result;
    }
};
// Logic-3
class Solution {
public:
    int binarySearch(vector<int>& row)
    {
        int low=0,high=row.size();
        while(low<high)
        {
            int mid=low+(high-low)/2;
            if(row[mid]==1)
                low=mid+1;
            else
                high=mid;
        }
        return low;
    }
    vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
        int r=mat.size(),c=mat[0].size();
        priority_queue<pair<int,int> > q;
        vector<int> result;
        for(int i=0;i<r;i++)
        {
            int count=binarySearch(mat[i]);
            q.push({count,i});
            if(q.size()>k)
                q.pop();
        }
        while(q.size())
        {
            int ele=q.top().second;
            result.push_back(ele);
            q.pop();
        }
        reverse(result.begin(),result.end());
        return result;
    }
};
int main()
{
    // https://leetcode.com/problems/the-k-weakest-rows-in-a-matrix/
    return 0;
}

File: Applied Course/4.Problem Solving/1.Problems on Array/24.The K Weakest Rows in a Matrix.py
# https://leetcode.com/problems/the-k-weakest-rows-in-a-matrix/
# Logic-1
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        counts=[(sum(row),i) for i,row in enumerate(mat)]
        counts.sort()
        return [counts[i][1] for i in range(k)]
# Logic-2
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        def binary_search(row):
            low=0
            high=len(row)
            while low<high:
                mid=low+(high-low)//2
                if row[mid]==1:
                    low=mid+1
                else:
                    high=mid
            return low
        counts=[(binary_search(row),i) for i,row in enumerate(mat)]
        counts.sort()
        return [counts[i][1] for i in range(k)]
# Logic-3
class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        m = len(mat)
        n = len(mat[0])

        def binary_search(row):
            low = 0
            high = n
            while low < high:
                mid = low + (high - low) // 2
                if row[mid] == 1:
                    low = mid + 1
                else:
                    high = mid
            return low

        # Calculate the strength of each row using binary search.
        # Put the strength/index pairs into a priority queue.
        pq = []
        for i, row in enumerate(mat):
            strength = binary_search(row)
            entry = (-strength, -i)
            if len(pq) < k or entry > pq[0]:
                heapq.heappush(pq, entry)
            if len(pq) > k:
                heapq.heappop(pq)

        # Pull out and return the indexes of the smallest k entries.
        # Don't forget to convert them back to positive numbers!
        indexes = []
        while pq:
            strength, i = heapq.heappop(pq)
            indexes.append(-i)

        # Reverse, as the indexes are around the wrong way.
        indexes = indexes[::-1]

        return indexes